mdio/mdioblock.1
.TH mdioblock 1 "November 2013" "open-plc-utils-0.0.3" "Qualcomm Atheros Open Powerline Toolkit"

.SH NAME
mdioblock - Qualcomm Atheros Clause 22 MDIO Instruction Block Assembler

.SH SYNOPSIS
.BR mdioblock
.RI [ options ]
.IR file
.RI [ file ]
.RI [ ... ]
.RI [> stdout ]

.PP
.BR mdioblock
.RI [ options ]
.RI [< stdin ] 
.RI [> stdout ]

.SH DESCRIPTION
Assemble an Atheros MDIO instruction block from hexadecimal register instructions.
MDIO instruction blocks are written to Atheros powerline devices as modules using the \fBVS_MOD_OP\fR message type.
Alternately, program \fBint6kmod\fR can be used.

.PP 
Atheros MDIO modules consists of a 16-bit header followed by one or more MDIO communication frames with register masks and padded to the nearest multiple of 32-bits for download purposes.
MDIO instructions are specified here using \fB802.3 Clause 22\fR (or \fIphy\fR, \fIreg\fR, \fIdata\fR and \fImask\fR) format.
See program \fBmdioblock2\fR for a similar program that accepts instructions in \fB802.3ae Clause 45\fR (or \fIaddress\fR, \fIdata\fR and \fImask\fR) format.

.PP
This program is part of the Qualcomm Atheros Powerline Toolkit.
See the \fBAMP\fR man page for an overview and installation instructions.

.SH OPTIONS

.TP
.RB - q
Quiet mode.
Suppresses error messages caused by instruction syntax errors.

.TP
.RB - v
Verbose mode.
Prints instructions on stderr as they are processed.

.TP
.RB - ? ,-- help   
Displays program help information on stderr.
This option takes precedence over all other options on the command line except version information.

.TP
.RB - ! ,-- version
Displays program version information on stderr.
This option takes precedence over all other options on the command line except help information.
Use this option when sending screen dumps to Atheros technical staff.

.SH ARGUMENTS

.TP
.IR file
A text files contining one or more generic MDIO register intructions consisting of a \fBphy\fR, \fBreg\fR, \fBdata\fR and \fBmask\fR fields.
Fields are seperated by whitespace and instructions are terminated with semicolon.
Multiple instructions may appear on a single line or a single instruction can span multiple lines.
Script-style comments are permitted between instructions but not between fields.

.SH REFERENCES
See the Qualcomm Atheros \fIAR7x00 Firmware Technical Reference Manual\fR section on the \fIModule Operation\fR message type for more information on MDIO Block format.

.SH DISCLAIMER
Atheros HomePlug AV Vendor Specific Management Message structure and content is proprietary to Qualcomm Atheros, Ocala FL USA.
Consequently, public information may not be available.
Qualcomm Atheros reserves the right to modify message structure and content in future firmware releases without any obligation to notify or compensate users of this program.

.SH EXAMPLES
The following command reads register instructions from file \fBmdio.txt\fR and writes them as an MDIO program block to file \fBmdio.bin\fR.
In this example, the program acts as a filter.

.PP
   # mdioblock < mdio.txt > mdio.bin

.PP
The following command reads register instructions from files \fBphysetup.dat\fR, \fBflowctl.dat\fR and \fBmisc.ins\fR and write them all as one MDIO block to file \fBmdio.bin\fR.
The previous contents of file \fBmdio.bin\fR is overwritten.

.PP
   # mdioblock physetup.dat flowctl.dat misc.ins > mdio.bin

.PP
An example register instruction file might look like this.
The first two lines consist of one instruction each.
An instruction has \fIphy\fR, \fIreg\fR, \fIdata\fR and \fImask\fR fields.
Instruction fields are separated by whitespace.
Instructions are separated by a semicolon.
Multiple instruction may appear on one line and one instruction can span multiple lines.
Script-style comments are permitted between instructions but not between instruction fields.

.PP
   18 00 0000 FFFF;
   10 16 003F FFFF;
   
   # multiple instructions on one line;
   
   10 17 7E3F FFFF; 18 00 0000 FFFF;
   
   # one instruction on multiple lines;
   
   10 
   
   02 
   
   0700 
   
   FFFF;
   
   10 03 0000 FFFF;
   18 00 0000 FFFF;
   14 00 007D FFFF;
   14 01 0000 FFFF; 

.PP
A 16-bit hexdump of the MDIO block will look like this on little-endian hosts.
The block consists of a 16-bit header followed by a variable number of 16-bit triplets.
The header indicates the number of triplets that follow.
Each triplet specifies a 16-bit operation, a 16-bit data value and a 16-bit data mask.
The application of these fields is beyond the scope of this man page.
Consult the IEEE standards for more information.

.PP
   # hexdump mdio.bin
   0000000 0241 8185 0000 ffff ad05 003f ffff af05
   0000010 7e3f ffff 8185 0000 ffff 8505 0700 ffff
   0000020 8705 0000 ffff 8185 0000 ffff 8145 007d
   0000030 ffff 8345 0000 ffff                    

.PP
An 8-bit dump of the MDIO block will look like this on all hosts since the MDIO block is always distributed in little-endian format.

.PP
   # hexdmp mdio.bin
   00000000  41 02 85 81 00 00 FF FF 05 AD 3F 00 FF FF 05 AF  A.........?.....
   00000010  3F 7E FF FF 85 81 00 00 FF FF 05 85 00 07 FF FF  ?~..............
   00000020  05 87 00 00 FF FF 85 81 00 00 FF FF 45 81 7D 00  ............E.}.
   00000030  FF FF 45 83 00 00 FF FF                          ..E.....

.SH SEE ALSO
.BR amp ( 1 ),
.BR int6kmod ( 1 ),
.BR int6kmdio ( 1 ),
.BR int6kmdio2 ( 1 ),
.BR mdioblock2 ( 1 ),
.BR mdiodump ( 1 ),
.BR mdiogen ( 1 )

.SH CREDITS
Charles Maier



CHANGES
open-plc-utils-0.0.6
---------------

1. Programs mac2pw and mac2pwd implement a new method of generating device passwords. The old method is supported but is not the default. 

open-plc-utils-0.0.5
---------------

open-plc-utils-0.0.4
---------------

open-plc-utils-0.0.3
---------------

1. Added header VisualStudioNET/include/inttype.h to define print specifier constant PRId64 for Windows systems.

2. Replaced conditional compiler code around print specifiers "lld" and "I64d" and replaced single specifier PRId64 as defined by the system.

3. Renamed man pages ending in .7 with .1 to reflect that fact that they will now be installed man1 instead of man7.

4. Replaced const HOMEPLUG_MTYPE with constant ETH_P_HPAV throughout.

5. Correct endianess of FIELD_ETH_TYPE in module coqos_add.c.

6. Updated chipset codes and implemented table-driven chipset detection algorithm.

7. Moved TLVNode structure definition from nvm/nvm.h to tools/tlv.h to prepare for exanded use of TLV structures in future MMEs.


open-plc-utils-0.0.2
---------------

1. Removed references to private vendor-specific MMEs

2. Removed unused archive targets in primary Makefile.

3. Added .gitignore to subdirectories where programs and other objects are compiled.

4. Added conditional __UCLIBC__ compile statement to ARCPPrint.c to account for variable argument definition error during compile; 

5. Added 'ignore' target to cascading Makefiles to re-write local .gitignore files.

6. Corrected email address for Florian Fainelli in module ARPCPrint.

7. Added new HomePlug AV SLAC Protocol MMEs definitions to homeplug.h.

8. Modified makefile 'ignore' target to use tr instead of sed because some sed implementations do not accept '\n'.



mdio/mdioblock.c
/*====================================================================*
 *
 *   Copyright (c) 2013 Qualcomm Atheros, Inc.
 *
 *   All rights reserved.
 *
 *   Redistribution and use in source and binary forms, with or
 *   without modification, are permitted (subject to the limitations
 *   in the disclaimer below) provided that the following conditions
 *   are met:
 *
 *   * Redistributions of source code must retain the above copyright
 *     notice, this list of conditions and the following disclaimer.
 *
 *   * Redistributions in binary form must reproduce the above
 *     copyright notice, this list of conditions and the following
 *     disclaimer in the documentation and/or other materials
 *     provided with the distribution.
 *
 *   * Neither the name of Qualcomm Atheros nor the names of
 *     its contributors may be used to endorse or promote products
 *     derived from this software without specific prior written
 *     permission.
 *
 *   NO EXPRESS OR IMPLIED LICENSES TO ANY PARTY'S PATENT RIGHTS ARE
 *   GRANTED BY THIS LICENSE. THIS SOFTWARE IS PROVIDED BY THE
 *   COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR
 *   IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 *   WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 *   PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER
 *   OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 *   SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 *   NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 *   LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 *   HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 *   CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
 *   OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 *   SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 *--------------------------------------------------------------------*/

/*====================================================================*"
 *
 *   mdioblock.c
 *
 *   Contributor(s):
 *      Charles Maier
 *
 *--------------------------------------------------------------------*/

/*====================================================================*
 *   system header files;
 *--------------------------------------------------------------------*/

#include <stdio.h>
#include <ctype.h>
#include <unistd.h>
#include <fcntl.h>
#include <sys/stat.h>

/*====================================================================*
 *   custom header files;
 *--------------------------------------------------------------------*/

#include "../tools/getoptv.h"
#include "../tools/endian.h"
#include "../tools/number.h"
#include "../tools/chars.h"
#include "../tools/error.h"
#include "../tools/flags.h"
#include "../mdio/mdio.h"

/*====================================================================*
 *   custom source files;
 *--------------------------------------------------------------------*/

#ifndef MAKEFILE
#include "../tools/getoptv.c"
#include "../tools/putoptv.c"
#include "../tools/version.c"
#include "../tools/error.c"
#include "../tools/todigit.c"
#endif

/*====================================================================*
 *   program variables;
 *--------------------------------------------------------------------*/

static signed c;
static unsigned row = 1;
static unsigned col = 1;
static uint16_t count = 0;
static uint16_t instr = 0;
static uint16_t addr;
static uint16_t data;
static uint16_t mask;
static uint16_t phy;
static uint16_t reg;

/*====================================================================*
 *
 *   signed mygetc ();
 *
 *   fetch next character from stdin and update the file cursor;
 *
 *--------------------------------------------------------------------*/

static signed mygetc ()

{
	signed c = getc (stdin);
	if (c == '\n')
	{
		row++;
		col = 0;
	}
	else if (~c)
	{
		col++;
	}
	return (c);
}


/*====================================================================*
 *
 *   uint16_t integer (unsigned radix);
 *
 *
 *--------------------------------------------------------------------*/

static uint16_t integer (unsigned radix)

{
	extern signed c;
	uint16_t value = 0;
	unsigned digit = 0;
	while ((digit = todigit (c)) < radix)
	{
		value *= radix;
		value += digit;
		c = mygetc ();
	}
	while (isspace (c))
	{
		c = mygetc ();
	}
	return (value);
}


/*====================================================================*
 *
 *   void assemble (flag_t flags);
 *
 *   read stdin and write stdout; convert hexadecimal input to binary
 *   MDIO register instructions;
 *
 *   the input file consists of zero or more hexadecimal instructions
 *   consisting of a phy address, register address, register data and
 *   register mask; instructions are terminated with semicolon; fields
 *   are separated by white space; scriptstyle comments are permitted
 *   between instructions but not between instruction fields;
 *
 *   the output file will consist of one 16-bit program header plus
 *   one 16-bit MDIO instructions for each input instruction; the
 *   output is padded to the nearest multiple of 32-bits;
 *
 *--------------------------------------------------------------------*/

static void assemble (flag_t flags)

{
	extern signed c;
	c = mygetc ();
	while (c != EOF)
	{
		if (isspace (c))
		{
			do
			{
				c = mygetc ();
			}
			while (isspace (c));
			continue;
		}
		if ((c == '#') || (c == ';'))
		{
			do
			{
				c = mygetc ();
			}
			while (nobreak (c));
			continue;
		}
		phy = integer (16);
		reg = integer (16);
		data = integer (16);
		mask = integer (16);
		instr = MDIO16_INSTR (1, 1, phy, reg, 2);
		write (STDOUT_FILENO, &instr, sizeof (instr));
		data = HTOLE16 (data & 0xFFFF);
		write (STDOUT_FILENO, &data, sizeof (data));
		mask = HTOLE16 (mask & 0xFFFF);
		write (STDOUT_FILENO, &mask, sizeof (mask));
		count++;
		if (_anyset (flags, MDIO_VERBOSE))
		{
			fprintf (stderr, "INSTR=0x%04X DATA=0x%04X MASK=0x%04X\n", instr, data, mask);
		}
		if ((c == ';') || (c == EOF))
		{
			c = mygetc ();
			continue;
		}
		if (_allclr (flags, MDIO_SILENCE))
		{
			error (1, 0, "Illegal character or missing terminator: line %d col %d", row, col);
		}
	}
	return;
}


/*====================================================================*
 *
 *   int main (int argc, const char * argv []);
 *
 *
 *
 *--------------------------------------------------------------------*/

int main (int argc, const char * argv [])

{
	static const char * optv [] =
	{
		"qv",
		"file [ file ] [ ...] > file",
		"Atheros Clause 22 MDIO Instruction Block Assembler",
		"q\tquiet mode",
		"v\tverbose mode",
		(const char *) (0)
	};
	flag_t flags = (flag_t)(0);
	optind = 1;
	while ((c = getoptv (argc, argv, optv)) != -1)
	{
		switch (c)
		{
		case 'q':
			_setbits (flags, MDIO_SILENCE);
			break;
		case 'v':
			_setbits (flags, MDIO_VERBOSE);
			break;
		default:
			break;
		}
	}
	argc -= optind;
	argv += optind;
	if (isatty (STDOUT_FILENO))
	{
		error (1, ECANCELED, "stdout must be a file or pipe");
	}

#if defined (WIN32)

	setmode (STDOUT_FILENO, O_BINARY);

#endif

	instr = MDIO16_START (1, 0, count);
	write (STDOUT_FILENO, &instr, sizeof (instr));
	if (!argc)
	{
		assemble (flags);
	}
	while ((argc) && (* argv))
	{
		if (!freopen (* argv, "rb", stdin))
		{
			error (1, errno, "%s", * argv);
		}
		assemble (flags);
		argc--;
		argv++;
	}
	instr = MDIO16_START (1, 0, count);
	addr = count * sizeof (instr) + sizeof (instr);
	if ((addr % sizeof (uint32_t)))
	{
		uint32_t pad = 0;
		write (STDOUT_FILENO, &pad, sizeof (pad) - addr % sizeof (pad));
	}
	if (!lseek (STDOUT_FILENO, 0, SEEK_SET))
	{
		write (STDOUT_FILENO, &instr, sizeof (instr));
	}
	if (_anyset (flags, MDIO_VERBOSE))
	{
		fprintf (stderr, "%d instructions\n", count);
	}
	return (0);
}

